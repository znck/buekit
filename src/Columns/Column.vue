<script>
import PropTypes, { normalize } from '@znck/prop-types'
import { styleResolver, style } from '../utils'
import { devices, additionalDevices, columns as sizes } from '../shared'
import { createTag } from '../mixins'

const sizeType = (name, extras = []) => PropTypes.oneOfType(
  PropTypes.number.validate(v => 1 <= v && v <= 12),
  PropTypes.string.validate(v => (x => 1 <= x && x <= 12)(Number(x))),
  PropTypes.oneOf(sizes)
).modifiers(name, [].concat(devices, extras))

let normalizer
const props = (props, _) => {
  normalizer = normalizer || normalize(props, (name, value) => {
      switch (name) {
        case 'size': return [_(`is-${value}`)]
        case 'offset': return [_(`is-${name}-${value}`)]
        case 'narrow': return value && [_(`is-${name}`)]
        default: return value
      }
    }, (name, suffix, value) => {
      switch (name) {
        case 'size': return _(`is-${value}-${suffix}`)
        case 'narrow': return value && _(`is-${name}-${suffix}`)
        case 'offset': return _(`is-${name}-${value}-${suffix}`)
      }
    }
  )

  return normalizer(props)
}

export default {
  name: 'Column',
  functional: true,
  mixins: [createTag()],
  props: {
    /**
     * Take only the space it needs.
     * @name narrow
     * @version 0.0.0
     * @since Version 0.0.0
     * @modifiers mobile tablet desktop
     */
    ...PropTypes.bool.value(false).modifiers('narrow', devices),
    /**
     * Horizontal space before a column element.
     * @name offset
     * @version 0.0.0
     * @since Version 0.0.0
     * @modifiers mobile tablet desktop
     */
    ...sizeType('offset'),
    /**
     * Width of column element in 12 point grid system.
     * @name size
     * @version 0.0.0
     * @since Version 0.0.0
     * @modifiers mobile tablet desktop
     */
    ...sizeType('size', additionalDevices)
  },

  render (h, ctx) {
    const _ = styleResolver(ctx.$style)
    const { tag, ...rest } = props(ctx.props, _)

    return style(h(tag, ctx.data, ctx.children), _('column'), Object.values(rest))
  }
}
</script>

<style lang="scss" src="./columns.scss" module></style>
