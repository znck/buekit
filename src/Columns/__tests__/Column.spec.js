import {shallow} from '@vue/test-utils'
import {createRenderer} from 'vue-server-renderer'

import Column from '../Column.vue'

const renderer = createRenderer()

function snapshots(options) {
	options.forEach(
    option => it('matches snapshot: ' + JSON.stringify(option), () => {
    const wrapper = shallow(Column, {
      context: {props: option},
      slots: {
        default: '<p>content</p>'
      }
    })

    expect(wrapper.contains('p')).toBe(true)

    renderer.renderToString(wrapper.vm, (err, str) => {
        if (err) {
          throw new Error(err)
        }

        expect(str).toMatchSnapshot()
      })
    })
  )
}

describe('Column.vue', () => {
	snapshots([
    {},
    {narrow: true},
    {'narrow.mobile': true},
    {'narrow.mobile': true, 'narrow.desktop': true},
    {offset: 'half'},
    {offset: 'half', 'offset$mobile': 3},
    {offset: 2},
    {size: 'half'},
    {size: 'half', 'size$mobile': 12},
    {size: 3}
  ])

  it(`size props`, () => {
    const wrapper = shallow(Column, {
      context: {
        props: { size: 'half', 'size$mobile': 12 },
      }
    })

    expect(wrapper.classes()).toEqual(['column', 'is-half', 'is-12-mobile'])
  })
})
