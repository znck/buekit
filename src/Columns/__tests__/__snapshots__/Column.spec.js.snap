// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column.vue matches snapshot: {"narrow":true} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-narrow\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"narrow.mobile":true,"narrow.desktop":true} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"narrow.mobile":true} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"offset":"half","offset$mobile":3} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-offset-half is-offset-3-mobile\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"offset":"half"} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-offset-half\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"offset":2} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-offset-2\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"size":"half","size$mobile":12} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-half is-12-mobile\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"size":"half"} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-half\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {"size":3} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column is-3\\"><p>content</p></div>"`;

exports[`Column.vue matches snapshot: {} 1`] = `"<div data-server-rendered=\\"true\\" class=\\"column\\"><p>content</p></div>"`;
